### DISCUSSION TITLE
Redacting all images

### DISCUSSION THREAD

**Discussion by northloopforensics** [pymupdf/PyMuPDF#914] [Looking for help] [answered]
First of all. This tool is incredible and the support is unreal.  Jorj, I am learning a lot just from reading the exchanges here. Thanks

I would like to redact every image in a PDF.  My attempt is:
`    

    doc = fitz.open("D:\\PYTHON WORK\\RegistryExplorerManual.pdf")

    for pageNumber, page in enumerate(doc.pages(), start=1):  #get page numbers for pdf
        for imgNumber, img in enumerate(page.getImageList(), start=1):#get image numbers for each page
            print(img[7])         
            box = page.getImageBbox(img[7])  #img[7] provides the image name
        
        # colors for redaction box
            yellow = (1, 1, 0)
            black = (0, 0, 0)
        
            page.addRedactAnnot(box, "REDACTED",align=fitz.TEXT_ALIGN_CENTER, fill=black, text_color=yellow)

    page.apply_redactions()

    doc.save("D:\\PYTHON WORK\\Redacted-Version.pdf", garbage=3, deflate=True)`

While this does not get me the desired effect, the output file has a red cross over every image I would like to redact.  So the tool is smarter than I am.  I know this is due to the page.apply_redactions() not being nested correctly, but can I leverage this to redact the images somehow?

**Comment by JorjMcKie** [comment, marked as answer]
First of all: thanks a lot for the compliments - very rewarding indeed!

When you say "Redact the images", then you mean delete them and put a textbox there instead, right?

1. From from can be seen in your snippet, the `page.apply_redactions()` is not indented correctly. As it is placed, it will only redact the last page of the document. Maybe that one didn't even have an image?
2. You should also check, whether an image **_reference_** found in `page.get_page_images()` really is an existing image on the page. Need not be the case because of how your PDF may be designed. The returned bbox will be "infinite" then.
2. Redacting the images per page should be indented one more level. So, when a page has 5 images, you should add 5 redaction annotations, and **then, in one go** redact the page - much more efficient.
2. You should add the `images` parameter `page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)`. This will prevent the MuPDF C-function from even trying to **modify** any images - which it does for every image that somehow overlaps any redaction. Instead it will remove overlapping images. A **_lot_** more efficient.

**Reply by northloopforensics** [reply]
Beautiful! Thank you for the quick response and your hard work on this tool!

![image](https://user-images.githubusercontent.com/73806121/108600920-634cd880-735f-11eb-808a-3a9aaf67b5cc.png)
